universal_transfer_operator/src/universal_transfer_operator/data_providers/filesystem/sftp.py

```diff
@@ -0,0 +1,143 @@
+
+    @property
+    def paths(self) -> list[str]:
+        """Resolve SFTP file paths with prefix"""
+        url = urlparse(self.dataset.path)
+        uri = self.get_uri()
+        full_paths = []
+        prefixes = self.hook.get_tree_map(url.netloc, prefix=url.netloc + url.path)
+        for keys in prefixes:
+            if len(keys) > 0:
+                full_paths.extend(keys)
+        # paths = ["/" + path for path in full_paths]
+        paths = [uri + "/" + path for path in full_paths]
+        return paths
+
+    @property
+    def transport_params(self) -> dict:
+        """get SFTP credentials for storage"""
+        client = self.hook.get_connection(self.dataset.conn_id)
+        extra_options = client.extra_dejson
+        if "key_file" in extra_options:
+            key_file = extra_options.get("key_file")
+            return {"connect_kwargs": {"key_filename": key_file}}
+        elif client.password:
+            return {"connect_kwargs": {"password": client.password}}
+        raise ValueError("SFTP credentials are not set in the connection.")
+
+    def get_uri(self):
+        client = self.hook.get_connection(self.dataset.conn_id)
+        return client.get_uri()
+
+    @staticmethod
+    def get_url_netloc(complete_url, dst_url):
+        keys = ["username", "password", "hostname", "port"]
+        final_vals = {}
+        for key in keys:
+            val = dst_url.__getattribute__(key)
+            if val is None:
+                val = complete_url.__getattribute__(key)
+            final_vals[key] = val
+
+        hostname = final_vals["hostname"]
+        complete_url_hostname = complete_url.__getattribute__("hostname")
+        if complete_url_hostname:
+            hostname = complete_url_hostname
+
+        final_vals["hostname"] = hostname
+
+        return "{username}:{password}@{hostname}:{port}".format(**final_vals)
+
+    @staticmethod
+    def get_url_path(complete_url, dst_url, src_url):
+        path = dst_url.path if dst_url.__getattribute__("path") else src_url.path
+        complete_url_hostname = complete_url.__getattribute__("hostname")
+        if complete_url_hostname:
+            path = dst_url.hostname + path
+        return path
+
+    def get_complete_url(self, dst_url, src_url) -> str:
+        """
+        Get complete url with host, port, credentials if they are not provided in the `dst_url`
+        """
+        complete_url = urlparse(self.get_uri())
+        _dst_url = urlparse(dst_url)
+        _src_url = urlparse(src_url)
+
+        netloc = self.get_url_netloc(complete_url=complete_url, dst_url=_dst_url)
+        path = self.get_url_path(complete_url=complete_url, dst_url=_dst_url, src_url=_src_url)
+
+        final_url = _dst_url._replace(netloc=netloc, path=path)
+
+        return urlunparse(final_url)
+
+    def write_using_smart_open(self, source_ref: FileStream):
+        """Write the source data from remote object i/o buffer to the dataset using smart open"""
```

Comment:
please add description of parameters and return value and add return value to types

Reply:
Done.